atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java

```diff
@@ -0,0 +1,188 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.cassandra.backup;
+
+import com.datastax.driver.core.Host;
+import com.datastax.driver.core.Metadata;
+import com.datastax.driver.core.Token;
+import com.datastax.driver.core.TokenRange;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import com.palantir.logsafe.exceptions.SafeIllegalStateException;
+import java.net.InetSocketAddress;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.SortedMap;
+import one.util.streamex.EntryStream;
+import one.util.streamex.StreamEx;
+
+public final class ClusterMetadataUtils {
+    private ClusterMetadataUtils() {
+        // util class
+    }
+
+    /**
+     * Returns a mapping of Node to token ranges its host contains, where every partition key in the specified
+     * list is present in one of the token ranges on one host.
+     *
+     * <p>For example there is a cluster that has keyspace K with a simple replication strategy and replication factor
+     * 2, the ring of the cluster has a range of (0, 3], and there are 3 nodes A, B, and C, where:
+     *
+     * <ul>
+     *   <li>A has replicas for K for token ranges (1, 2] and (2, 3]
+     *   <li>B has replicas for K for token ranges (2, 3] and (0, 1]
+     *   <li>C has replicas for K for token ranges (0, 1] and (1, 2]
+     * </ul>
+     *
+     * <p>The value returned by calling getTokenMapping([A, B, C], metadata, K, [0.25, 0.5, 2.5]) could be: <br>
+     * { A : set[ range(2, 3] ] B : set[ range(0, 1] ] }
+     *
+     * <p>Another possible return value is: <br>
+     * { B : set[ range(0, 1], range(2, 3] ] }
+     *
+     * <p>The node chosen for each token range can be any replica, so it is non-deterministic.
+     *
+     * @param nodeSet The Cassandra nodes whose replicas to check
+     * @param metadata The Datastax driver metadata from the Cassandra cluster
+     * @param keyspace The keyspace containing the partition keys we want to map
+     * @param partitionKeyTokens The Cassandra tokens for the partition keys we want to map
+     * @return Mapping of Node to token ranges its host contains, where every partition key in the specified
+     * list is present in one of the token ranges on one host
+     */
+    @SuppressWarnings("ReverseDnsLookup")
+    public static Map<InetSocketAddress, Set<TokenRange>> getTokenMapping(
+            Collection<InetSocketAddress> nodeSet, Metadata metadata, String keyspace, Set<Token> partitionKeyTokens) {
```

Comment:
So the tokens returns aren't the same - they're a proper subset of the token map, except in the edge case where the end point of each range in the token ring has been hit.

I've added a test that shows that the ranges returned are of the form that we expect.

Reply:
It is probably worth it to use this code, as we do need to migrate to CQL eventually anyway

Reply:
Overall this class is a faithful representation of the internal product, but it can definitely benefit from a moderate refactor as outlined below. Also, please add unit tests!